caikit_nlp/modules/text_generation/fine_tuning.py

```diff
@@ -0,0 +1,262 @@
+# Copyright The Caikit Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+# Third Party
+from torch.utils.data import IterableDataset
+from transformers import (
+    AutoConfig,
+    AutoTokenizer,
+    DataCollatorForSeq2Seq,
+    Seq2SeqTrainer,
+    Seq2SeqTrainingArguments,
+    Trainer,
+)
+
+# First Party
+from caikit.core.data_model import DataStream
+from caikit.core.modules import ModuleBase, module
+from caikit.core.toolkit import error_handler, wip_decorator
+import alog
+
+# Local
+from ...data_model import GeneratedResult, GenerationTrainRecord
+from ...toolkit.data_stream_wrapper import SimpleIterableStreamWrapper
+from ...toolkit.data_type_utils import get_torch_dtype
+from .text_generation_task import TextGenerationTask
+
+log = alog.use_channel("TXT_GEN")
+error = error_handler.get(log)
+
+
+# pylint: disable=too-many-lines,too-many-instance-attributes
+@module(
+    id="28a81449-32ce-4be3-b688-545bde68f738",
+    name="Text Generation",
+    version="0.1.0",
+    task=TextGenerationTask,
+)
+@wip_decorator.work_in_progress(
+    category=wip_decorator.WipCategory.WIP, action=wip_decorator.Action.ERROR
+)
+class FineTuning(ModuleBase):
+    """Module to provide fine-tuning support for text generation task"""
+
+    def __init__(self, tokenizer, model):
+        super().__init__()
+
+        self.tokenizer = tokenizer
+        # NOTE: self.model here can also be HF trainer. This is because
+        # if we have just trained the model then the models weights might be
+        # available in different devices (and configuration), depending on
+        # how it was trained. For now (July 10, 2023), we are not trying to
+        # extract the model out from trainer itself, since that would require
+        # us to essentially save it or reconstruct it to do normal inferring.
+        self.model = model
+
+    @classmethod
+    def train(
+        cls,
+        base_model: str,  # TODO: Union[str, PretrainedModelBase]
```

Comment:
Somewhat unrelated - I can't remember the exact reason unions were disallowed in our type hints for a little while, maybe something for caikit runtime type hint inspection and unions being unsupported at the time. I think that some work has been going on around that recently though - do you know if we can use them yet?

Reply:
hmm. honestly not sure, I guess certainly something we could try, but there is a lot of things influx in data models, so we might want things to be more stable and then make these changes